### DISCUSSION TITLE
hostfxr_run_app & hostfxr_get_runtime_delegate

### DISCUSSION THREAD

**Discussion by wingchunninja** [dotnet/runtime#76358] [Q&A] [answered]
Hi everyone,

I have a .net6.0 application that I call from unmanaged code.
What works is using **hostfxr_get_runtime_delegate** and then calling functions from a particular component.

If I call only **hostfxr_get_runtime_delegate**, I have the problem that the Main/Startup classes from my .net application are not being called, and that leads to the point where crucial things that I use in the component class are not being initialized.

Is there a possibility to first call hostfxr_run_app and then **hostfxr_get_runtime_delegate**. What I have tested is that calling **hostfxr_run_app** works and initialized the app, but after that calling **hostfxr_get_runtime_delegate** on the same contextHandle results into me receiving an **EXC_BAD_ACCESS (code=1, address=0x986d41350c08)**.

Do you have some ideas how i can use a hybrid from both, initializing the app and after that calling the component?

**Comment by vitek-karas** [comment, marked as answer]
The `hostfxr_run_app` will only return once `Main` returns at which point the runtime shuts down. So using stuff after that is not possible.

There are two possible solutions if you need to have:
* Custom host - so that you control how the runtime is loaded
* Running managed app inside the process

1) You could just call `Main` as a normal method. This comes with some gotchas:
* Normally `Main` will not return until the app is done, so may want to do this on a new thread and let it sit there (depends on what it does)
* Some things will not be initialized the same way they would be with `hostfxr_run_app`. For example things like `Environment.GetCommandLineArgs` will not return any value. It's unlikely this would be a problem, but it can.

2) A cleaner solution is to use `hostfxr_run_app` and then:
* The managed code calls back into native - typically via PInvokes
* It passes a native function pointer to some managed method which the native code stores
* Native code then can call into managed using the native function pointer at any point
(This is roughly what ASP.NET does when being hosted inside IIS)

**Reply by wingchunninja** [reply]
Thanks, i already did approach 1. But I think i will try the second one as well.
